In [3]:
#hide
options(jupyter.plot_scale=3)
options(repr.plot.width=6,repr.plot.height=6,repr.plot.res=300)
options(jupyter.rich_display=FALSE)
options(max.print=1000)

# (노트) 함수사용 고급

- branch: master
- badges: true
- comments: true
- author: 신록예찬
- categories: [R]

In [5]:
f<-function(x) x^2 
f

function(x) x^2

`-` formals(): 함수의 입력을 알려줌 

In [7]:
formals(f)

$x



`-` body(): 함수의 내용을 알려줌 

In [8]:
body(f)

x^2

`-` 함수가 정의된 환경을 알려줌 

In [9]:
environment(f)

<environment: R_GlobalEnv>

### 스코핑 

`-` 예제1

In [28]:
f<-function(){
    x<-1
    y<-2
    c(x,y)
} 

In [29]:
f()

[1] 1 2

`-` 예제2

In [30]:
x<-2 
f<- function(){
    y<-1 
    c(x,y)
}
f()

[1] 2 1

`-` 예제3

In [31]:
x<-1
f<-function(){
    y<-2
    g<-function(){
        z<-3
        c(x,y,z)
    }
    g()
}

In [32]:
f()

[1] 1 2 3

`-` 예제4

In [36]:
f <- function(x){
    y<-2 
    function() c(x,y)
}
g <- f(1)

In [37]:
g()

[1] 1 2

`-` 예제5

In [42]:
f <- function(x) x+1 
g <- function(){
    f <- function(x) x*2
    f(10)
}

In [45]:
g()

[1] 20

In [ ]:
f(10)

[1] 11

### 동적탐색  

`-` R은 함수가 생성될때가 아니라 실행될때 값을 탐색한다. 

In [52]:
f<-function() x 
x<-15

In [53]:
f()

[1] 15

### R의 비밀 

***존재하는 모든것은 오브젝트이다. 그리고 발생하는 모든것은 함수호출이다.***

`-` 예제1

In [58]:
f = function(x,y) (x+y)**2

In [59]:
f(10,5)

[1] 225

발생한것은 225이고 슬로건에 따르면 225는 어떠한 함수의 결과이다. (f라는 이름을 가진 함수의 결과이다. 당연한걸 설명하나?) 

`-` 예제2

In [60]:
10+20 

[1] 30

발생하는것은 30, 따라서 슬로건에 따르면 30은 어떠한 함수의 결과이다. 

In [61]:
`+`(10,20)

[1] 30

- 사실은 `+` 라는 이름의 함수가 있어서 이를 적용한 결과였던것임 

`-` 예제3

In [62]:
for (i in 1:2) print(i)

[1] 1
[1] 2


발생하는것은 1,2 인데 슬러건에 따르면 이것은 `for`라는 함수의 결과인가? 

In [64]:
`for`(i, 1:2, print(i))

[1] 1
[1] 2


- `for`라는 함수가 존재했고, 그 함수는 i, 1:2, print(i))를 입력을 받고있었음 

`-` 예제4

In [65]:
for (i in 1:2){
    i=i+1 
    print(i)
}

[1] 2
[1] 3


In [71]:
`for`(i ,1:2, {i=i+1 
               print(i)})

[1] 2
[1] 3


In [72]:
`for`(i ,1:2, {i=i+1;print(i)})

[1] 2
[1] 3


`-` 예제5